redash/handlers/dashboards.py

```diff
@@ -49,18 +77,45 @@ def get(self, dashboard_slug=None):
             response['public_url'] = url_for('redash.public_dashboard', token=api_key.api_key, org_slug=self.current_org.slug, _external=True)
             response['api_key'] = api_key.api_key
 
+        last_change = models.Change.get_latest(object_id=dashboard.id, object_type=models.Dashboard.__name__)
+        if last_change:
+            response['latest_version'] = last_change.id
+
         return response
 
     @require_permission('edit_dashboard')
     def post(self, dashboard_slug):
         dashboard_properties = request.get_json(force=True)
         # TODO: either convert all requests to use slugs or ids
         dashboard = models.Dashboard.get_by_id_and_org(dashboard_slug, self.current_org)
+
+        # check access permissions
+        if self.current_user.id != dashboard.user.id:
+            if not self.current_user.has_access(
+                    access_type=models.AccessPermission.ACCESS_TYPE_MODIFY,
+                    object_id=dashboard.id,
+                    object_type=models.Dashboard.__name__):
+                abort(403)
+
+        # Optimistic locking: figure out which user made the last
+        # change to this dashboard, and bail out if necessary
+        last_change = models.Change.get_latest(object_id=dashboard.id, object_type=models.Dashboard.__name__)
+        if last_change and 'latest_version' in dashboard_properties:
+            if last_change.id > dashboard_properties['latest_version']:
+                abort(409) # HTTP 'Conflict' status code
```

Comment:
This part of the handler has also been updated to work with the new `version` property in the model. Conceptually, it should provide the same functionality as the snippet that Charles had suggested.